server/taskCheckCost.go

```diff
@@ -0,0 +1,168 @@
+	} else {
+		runCostCheckForAccount(ctx, aa)
+	}
+	if err != nil {
+		logger.Error("Failed to check account cost.", map[string]interface{}{
+			"awsAccountId": aaId,
+			"error":        err.Error(),
+		})
+	}
+	return
+}
+
+func getCostFromExplorer(ctx context.Context, aa taws.AwsAccount, intervalBegin, currentMonthBeginning time.Time) (*costexplorer.GetCostAndUsageOutput, error) {
+	creds, err := taws.GetTemporaryCredentials(aa, "trackit-cost-check")
+	if err != nil {
+		return nil, err
+	}
+	sess := session.Must(session.NewSession(&aws.Config{
+		Credentials: creds,
+	}))
+	svc := costexplorer.New(sess)
+
+	explorerGranularity := "MONTHLY"
+	explorerIntervalStart := intervalBegin.Format("2006-01-02")
+	explorerIntervalEnd := currentMonthBeginning.Format("2006-01-02")
+	explorerMetrics := "UnblendedCost"
+	costExplorerInput := costexplorer.GetCostAndUsageInput{
+		Granularity: &explorerGranularity,
+		TimePeriod: &costexplorer.DateInterval{
+			Start: &explorerIntervalStart,
+			End:   &explorerIntervalEnd,
+		},
+		Metrics: []*string{&explorerMetrics},
+	}
+	return svc.GetCostAndUsage(&costExplorerInput)
+}
+
+func getCostFromES(ctx context.Context, aa taws.AwsAccount, intervalBegin, intervalEnd time.Time) (es.SimplifiedCostsDocument, int, error) {
+	accountList := []string{aa.AwsIdentity}
+	aggregationParams := []string{"month"}
+	indexList := []string{es.IndexNameForUserId(aa.UserId, s3.IndexPrefixLineItem)}
+	params := costs.EsQueryParams{
+		DateBegin:         intervalBegin,
+		DateEnd:           intervalEnd,
+		AccountList:       accountList,
+		IndexList:         indexList,
+		AggregationParams: aggregationParams,
+	}
+	return costs.MakeElasticSearchRequestAndParseIt(ctx, params)
+}
+
+// runCostCheckForAccount runs a cost check for an account
+func runCostCheckForAccount(ctx context.Context, aa taws.AwsAccount) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	now := time.Now().UTC()
+	currentMonthBeginning := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
+	intervalEnd := currentMonthBeginning.Add(time.Nanosecond * -1)
+	intervalBegin := currentMonthBeginning.AddDate(0, -6, 0)
+
+	explorerCosts, err := getCostFromExplorer(ctx, aa, intervalBegin, currentMonthBeginning)
+	if err != nil {
+		logger.Error("Failed to retrieve cost from the cost explorer", err.Error())
+		return
+	}
+	esCosts, _, err := getCostFromES(ctx, aa, intervalBegin, intervalEnd)
+	if err != nil {
+		logger.Error("Failed to retrieve cost from ES", err.Error())
+		return
+	}
+	if len(explorerCosts.ResultsByTime) != len(esCosts.Children) {
+		logger.Error("Number of month does not match for cost comparison", map[string]interface{}{
+			"explorerMonthNb": len(explorerCosts.ResultsByTime),
+			"esMonthNb":       len(esCosts.Children),
+		})
+		return
+	}
+	for i, esMonth := range esCosts.Children {
+		explorerMonthCostPtr := explorerCosts.ResultsByTime[i].Total["UnblendedCost"].Amount
```

Comment:
API does not guarantee any month order, and probably as well as AWS Explorer.

Reply:
From my tests they are in a good order. This is only a task to help us do some debug on the trackit account so it's not a big deal if the code is not perfect. Should I still spend time to do the modifications?

Reply:
Up to you, just to let you know it might be random.